File: topics/topics/array/problems/ContainerSystem.py
from collections import defaultdict
from sortedcontainers import SortedSet

class ContainerSystem:
    index_to_num = {}
    num_to_indexes = defaultdict(SortedSet)

    def change(self, index, number):
        if index in self.index_to_num:
            earlier_num = self.index_to_num[index]
            self.num_to_indexes[earlier_num].remove(index)
        if number not in self.num_to_indexes:
            self.num_to_indexes[number] = SortedSet()
        self.num_to_indexes[number].add(index)
        self.index_to_num[index] = number
        print(self.num_to_indexes)
        # print(self.index_to_num)

    def find(self, number):
        if number not in self.num_to_indexes or len(self.num_to_indexes[number]) == 0:
            return -1
        else:
            sorted_set = self.num_to_indexes[number]
            return (sorted_set[0])
            #return self.num_to_indexes[number]

input = [[10], [2, 10], [1, 10], [3, 10], [5, 10], [10], [1, 20], [10]]
cs = ContainerSystem()
for i in input:
    if len(i)==2:
        cs.change(i[0], i[1])
    else:
        print(cs.find(i[0]))

File: topics/topics/array/problems/ContainerSystem.java
package array.problems;

import java.util.*;

public class ContainerSystem {

    public static void main(String[] args) {
        ContainerSystem containerSystem = new ContainerSystem();
        containerSystem.designSystem();
    }

    private void designSystem() {

        Map<Integer, Integer> indexToNum = new HashMap<>();
        Map<Integer, SortedSet<Integer>> numberToIndexes = new HashMap<>();


        int[][] inputs = {{25, 50}, {56, 31}, {50}, {50}, {43}, {30, 50}, {31}, {43}, {25, 20}, {50}, {56, 43}, {68, 31}, {56, 31}, {20}, {43}, {25, 43}, {43}, {56, 31}, {54, 43}, {63, 43}};
        for (int[] input : inputs) {
            if (input.length == 2) {
                int index = input[0];
                int currNum = input[1];
                if (indexToNum.containsKey(index)) {
                    int earlierNum = indexToNum.get(index);
                    numberToIndexes.get(earlierNum).remove(index);
                }
                if (!numberToIndexes.containsKey(currNum)) numberToIndexes.put(currNum, new TreeSet<>());
                numberToIndexes.get(currNum).add(index);
                indexToNum.put(index, currNum);
            } else {
                if (numberToIndexes.get(input[0]) == null || numberToIndexes.get(input[0]).isEmpty()) System.out.println(-1);
                else System.out.println(numberToIndexes.get(input[0]).first());
            }
        }

    }

}
